pkg/requestreply/key_store.go

```diff
@@ -0,0 +1,252 @@
+	keyStores map[types.NamespacedName]*requestReplyAesKeyStore
+	done      chan bool
+	Logger    *zap.SugaredLogger
+}
+
+func (ks *AESKeyStore) AddAesKey(rrName types.NamespacedName, keyName string, keyValue []byte) {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+
+	if ks.keyStores == nil {
+		ks.keyStores = make(map[types.NamespacedName]*requestReplyAesKeyStore)
+	}
+
+	if ks.keyStores[rrName] == nil {
+		ks.keyStores[rrName] = &requestReplyAesKeyStore{}
+	}
+
+	ks.keyStores[rrName].addAesKey(keyName, keyValue)
+}
+
+func (ks *AESKeyStore) RemoveAesKey(rrName types.NamespacedName, keyName string) {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+
+	if ks.keyStores[rrName] == nil {
+		return
+	}
+
+	ks.keyStores[rrName].removeAesKey(keyName)
+}
+
+func (ks *AESKeyStore) GetLatestKey(rrName types.NamespacedName) ([]byte, bool) {
+	ks.lock.RLock()
+	defer ks.lock.RUnlock()
+
+	if ks.keyStores[rrName] == nil {
+		return nil, false
+	}
+
+	return ks.keyStores[rrName].getLatestKey(), true
+}
+
+func (ks *AESKeyStore) GetAllKeys(rrName types.NamespacedName) ([][]byte, bool) {
+	ks.lock.RLock()
+	defer ks.lock.RUnlock()
+
+	if ks.keyStores[rrName] == nil {
+		return nil, false
+	}
+
+	return ks.keyStores[rrName].getAllKeys(), true
+}
+
+func (ks *AESKeyStore) StopWatch() {
+	ks.done <- true
+	close(ks.done)
+}
+
+func (ks *AESKeyStore) WatchPath(basePath string) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create file watcher: %w", err)
+	}
+
+	if ks.done == nil {
+		ks.done = make(chan bool)
+	}
+
+	go func() {
+		defer watcher.Close()
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				ks.Logger.Debug("received fsnotify event", zap.Any("event", event))
+				if !ok {
+					return
+				}
+				ks.handleEvent(event)
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				ks.Logger.Error("file watcher encountered an error", zap.Error(err))
+			case done, ok := <-ks.done:
+				if done || !ok {
+					ks.Logger.Info("closing watcher...")
+					return
+				}
+			}
+		}
+
+	}()
+
+	ks.Logger.Infof("starting watch of files in %s", basePath)
+	watcher.Add(basePath)
+
+	ks.Logger.Infof("performing initial scan of files in %s", basePath)
+	return filepath.Walk(basePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.Mode().IsRegular() {
+			ks.processKeyFile(path)
+		}
+
+		return nil
+	})
+}
+
+func (ks *AESKeyStore) handleEvent(event fsnotify.Event) {
+	if event.Has(fsnotify.Create) {
+		info, err := os.Stat(event.Name)
+		if err != nil {
+			ks.Logger.Warnw("failed to stat file", zap.String("file", event.Name), zap.Error(err))
+			return
+		}
+
+		if info.Mode().IsRegular() {
+			ks.processKeyFile(event.Name)
+		}
+	}
+
+	if event.Has(fsnotify.Write) {
+		ks.processKeyFile(event.Name)
+	}
+
+	if event.Has(fsnotify.Remove) {
+		ks.removeKeyFile(event.Name)
+	}
+}
+
+func (ks *AESKeyStore) processKeyFile(path string) {
```

Comment:
these are called by the goroutine. Should the handle be atomic to prevent races when processing filesystem calls?